miniapps/shifted/extrapolate.cpp

```diff
@@ -0,0 +1,219 @@
+      std::cout << std::setprecision(12) << std::fixed
+                << text << norm << std::endl;
+   }
+}
+
+void PrintIntegral(int myid, ParGridFunction &g, std::string text)
+{
+   ConstantCoefficient zero(0.0);
+   double norm = g.ComputeL1Error(zero);
+   if (myid == 0)
+   {
+      std::cout << std::setprecision(12) << std::fixed
+                << text << norm << std::endl;
+   }
+}
+
+int main(int argc, char *argv[])
+{
+   // Initialize MPI.
+   MPI_Session mpi;
+   int myid = mpi.WorldRank();
+
+   // Parse command-line options.
+   const char *mesh_file = "../../data/inline-quad.mesh";
+   int rs_levels = 2;
+   Extrapolator::XtrapType ex_type   = Extrapolator::ASLAM;
+   AdvectionOper::AdvectionMode dg_mode = AdvectionOper::HO;
+   int ex_order = 1;
+   int order = 2;
+   double distance = 0.35;
+   bool vis_on = true;
+   int vis_steps_cnt = 50;
+
+   OptionsParser args(argc, argv);
+   args.AddOption(&mesh_file, "-m", "--mesh",
+                  "Mesh file to use.");
+   args.AddOption(&rs_levels, "-rs", "--refine-serial",
+                  "Number of times to refine the mesh uniformly in serial.");
+   args.AddOption((int*)&ex_type, "-et", "--extrap-type",
+                  "Extrapolation type: Aslam (0) or Bochkov (1).");
+   args.AddOption((int*)&dg_mode, "-dg", "--dg-mode",
+                  "DG advection mode: 0 - Standard High-Order,\n\t"
+                  "                   1 - Low-Order Upwind Diffusion.");
+   args.AddOption(&ex_order, "-eo", "--extrap-order",
+                  "Extrapolation order: 0/1/2 for constant/linear/quadratic.");
+   args.AddOption(&order, "-o", "--order",
+                  "Finite element order (polynomial degree) or -1 for"
+                  " isoparametric space.");
+   args.AddOption(&distance, "-d", "--distance",
+                  "Extrapolation distance.");
+   args.AddOption(&problem, "-p", "--problem",
+                  "0 - 2D circle,\n\t"
+                  "1 - 2D star");
+   args.AddOption(&vis_on, "-vis", "--visualization", "-no-vis",
+                  "--no-visualization",
+                  "Enable or disable GLVis visualization.");
+   args.AddOption(&vis_steps_cnt, "-vs", "--visualization-steps",
+                  "Visualize every n-th timestep.");
+   args.Parse();
+   if (!args.Good())
+   {
+      if (myid == 0) { args.PrintUsage(cout); }
+      return 1;
+   }
+   if (myid == 0) { args.PrintOptions(cout); }
+
+
+   // Refine the mesh and distribute.
+   Mesh mesh(mesh_file, 1, 1);
+   for (int lev = 0; lev < rs_levels; lev++) { mesh.UniformRefinement(); }
+   ParMesh pmesh(MPI_COMM_WORLD, mesh);
+   mesh.Clear();
+   const int dim = pmesh.Dimension();
+
+   // Input function.
+   L2_FECollection fec_L2(order, dim);
+   ParFiniteElementSpace pfes_L2(&pmesh, &fec_L2);
+   ParGridFunction u(&pfes_L2);
+   FunctionCoefficient u0_coeff(solution0);
+   u.ProjectCoefficient(u0_coeff);
+
+   // Extrapolate.
+   Extrapolator xtrap;
+   xtrap.xtrap_type     = ex_type;
```

Comment:
Can lines 177 - 181 be incorporated to the constructor of Extrapolator and those member variables be set to private?

Reply:
Matter of choice. This allows dynamic behavior of the object. I think it's safe to leave them public - the class is simple, and these variables only affect the behavior of `Extrapolate()`.